snuba/subscriptions/store.py

```diff
@@ -0,0 +1,104 @@
+import json
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import Collection
+
+from rediscluster import StrictRedisCluster
+
+from snuba.subscriptions.data import Subscription
+from snuba.utils.codecs import Codec
+
+
+@dataclass(frozen=True)
+class RedisPayload:
+    __slots__ = ["key", "value"]
+
+    key: str
+    value: bytes
+
+
+class SubscriptionCodec(Codec[RedisPayload, Subscription]):
+    def encode(self, value: Subscription) -> RedisPayload:
+        return RedisPayload(
+            value.id,
+            json.dumps(
+                {
+                    "id": value.id,
+                    "project_id": value.project_id,
+                    "conditions": value.conditions,
+                    "aggregations": value.aggregations,
+                    "time_window": int(value.time_window.total_seconds()),
+                    "resolution": int(value.resolution.total_seconds()),
+                }
+            ).encode("utf-8"),
+        )
+
+    def decode(self, value: RedisPayload) -> Subscription:
+        data = json.loads(value.value.decode("utf-8"))
+        return Subscription(
+            id=data["id"],
+            project_id=data["project_id"],
+            conditions=data["conditions"],
+            aggregations=data["aggregations"],
+            time_window=timedelta(seconds=data["time_window"]),
+            resolution=timedelta(seconds=data["resolution"]),
+        )
+
+
+class SubscriptionDoesNotExist(Exception):
+    pass
+
+
+class RedisSubscriptionStore:
+    """
+    A Redis backed store for Subscriptions. Stores subscriptions using
+    `SubscriptionCodec`. Each instance of the store operates on a partition of data,
+    defined by the `key` constructor param.
+    """
+
+    KEY_TEMPLATE = "subscriptions:{}"
```

Comment:
Not a strong opinion here, but I think it'd be a bit simpler/clearer to avoid any templating/formatting of the key passed to the constructor (removing this and `_key`.)

Reply:
To me it feels a little weird for the consumer of this class to be passing in the namespace. At least for the moment, I don't see why we'd have multiple namespaces.

Reply:
Yeah, I don't have a strong opinion in this, but I do think it sort of bakes in (or at least implies) the partitioning implementation detail into this class which doesn't really need to be aware of it. It'll obviously work either way though.